common/gesture.c

```diff
@@ -0,0 +1,352 @@
+		return false;
+	}
+
+	// Check that finger count matches
+	if (target->fingers != GESTURE_FINGERS_ANY &&
+		target->fingers != fingers) {
+		return false;
+	}
+
+	return true;
+}
+
+bool gesture_match(struct gesture *target, struct gesture *to_match, bool exact) {
+	// Check type and fingers
+	if(!gesture_check(target, to_match->type, to_match->fingers)) {
+		return false;
+	}
+
+	// Enforce exact matches ...
+	if(exact && target->directions != to_match->directions) {
+		return false;
+	}
+
+	// ... or ensure all target directions are matched
+	return (target->directions & to_match->directions) == target->directions;
+}
+
+bool gesture_equal(struct gesture *a, struct gesture *b) {
+	return a->type == b->type
+		&& a->fingers == b->fingers
+		&& a->directions == b->directions;
+}
+
+// Return count of set bits in directions bit field.
+static uint8_t gesture_directions_count(uint32_t directions) {
+	uint8_t count = 0;
+	for (; directions; directions >>= 1){
+		count += directions & 1;
+	}
+	return count;
+}
+
+// Compare direction bit count of two direction.
+static int8_t gesture_directions_compare(uint32_t a, uint32_t b) {
+	return gesture_directions_count(a) - gesture_directions_count(b);
+}
+
+// Compare two direction based on their direction bit count
+int8_t gesture_compare(struct gesture *a, struct gesture *b) {
+	return gesture_directions_compare(a->directions, b->directions);
+}
+
+void gesture_tracker_begin(struct gesture_tracker *tracker,
+		enum gesture_type type, uint8_t fingers) {
+	tracker->type = type;
+	tracker->fingers = fingers;
+
+	tracker->dx = 0.0;
+	tracker->dy = 0.0;
+	tracker->scale = 1.0;
+	tracker->rotation = 0.0;
+
+	sway_log(SWAY_DEBUG, "begin tracking %s:%u:? gesture",
+		gesture_type_string(type),
+		fingers);
+}
+
+bool gesture_tracker_check(struct gesture_tracker *tracker, enum gesture_type type) {
+	return tracker->type == type;
+}
+
+void gesture_tracker_update(struct gesture_tracker *tracker,
+		double dx, double dy, double scale, double rotation) {
+	if (tracker->type == GESTURE_TYPE_HOLD) {
+		sway_assert(false, "hold does not update.");
+		return;
+	}
+
+	tracker->dx += dx;
+	tracker->dy += dy;
+
+	if (tracker->type == GESTURE_TYPE_PINCH) {
+		tracker->scale = scale;
+		tracker->rotation += rotation;
+	}
+
+	sway_log(SWAY_DEBUG, "update tracking %s:%u:? gesture: %f %f %f %f",
+			gesture_type_string(tracker->type),
+			tracker->fingers,
+			tracker->dx, tracker->dy,
+			tracker->scale, tracker->rotation);
+}
+
+void gesture_tracker_cancel(struct gesture_tracker *tracker) {
+	sway_log(SWAY_DEBUG, "cancel tracking %s:%u:? gesture",
+			gesture_type_string(tracker->type), tracker->fingers);
+
+	tracker->type = GESTURE_TYPE_NONE;
+}
+
+struct gesture *gesture_tracker_end(struct gesture_tracker *tracker) {
+	struct gesture *result = calloc(1, sizeof(struct gesture));
+
+	// Ignore gesture under some threshold
+	// TODO: Make configurable
+	const double min_rotation = 5;
+	const double min_scale_delta = 0.1;
```

Comment:
@RPigott What do you think about these? I probably should make them configurable.

Reply:
New configuration entries will need to be maintained/supported forever because Sway's config file format is stable. I'd prefer hardcoding values for now, try to find better values if this is an issue, and only add a config option if it turns out we have no choice.

Reply:
I guess requiring at least a five degree rotation and at least a 10% scale change is reasonable.